### PR TITLE
[Catalog] Allow placeholder value to be any type

### PR DESCRIPTION
**Pull request by taras**
Related #13288

## Hey, I just made a Pull Request!

I figured out how to use PlaceholderProcessor to prevent duplication in templates. In the process, we discovered that the processor only allowed the value to be a string. This is too restrictive in cases when you want to use the placeholder; that's an arbitrary function that returns a value. 

1. I changed the logic to only ignore the placeholder when a resolver doesn't exist.
2. I added a TODO to make resolving this value recursive, but I believe this should be executed from the bottom up. Thoughts?
3. Added basic documentation to `writing-templates.md` to show how to create a placeholder for a template.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

## How to use this functionality

We can use the PlaceholderProcessor to create reusable portions of a template. A placeholder is a property on an entity object that starts with `$`. Backstage has some built in placeholders including `$text`, `$json` and `$yaml`. Each placeholder has a resolver. A resolver is an asyncronous function that receives the value assigned to the placeholder. A resolver is expected to return a promise that resolves to a value. The resolved value will replace the object where the placeholder was defined.

Let's say we want to reuse a portion of the template that asks user to specify a host name and we want to be able to add other fields to that form step.

```yaml
apiVersion: scaffolder.backstage.io/v1beta3
kind: Template
metadata:
  name: v1beta3-demo
  title: Microservice example
  description: scaffolder v1beta3 template demo
spec:
  owner: backstage/techdocs-core
  type: service
  parameters:
    - title: Specify a host
      parameters:
        name:
          title: Hostname
          type: string
          description: Specify host name
```

Our placeholder is going to be called `$specifyHostname`. We'd use it like this,

```yaml
apiVersion: scaffolder.backstage.io/v1beta3
kind: Template
metadata:
  name: v1beta3-demo
  title: Microservice example
  description: scaffolder v1beta3 template demo
spec:
  owner: backstage/techdocs-core
  type: service
  parameters:
    - $specifyHostname:
      domainExt:
        name: Domain extension
        type: string
        description: Specify domain extension like .com, .ca, .co.uk or something else.
```

The result after the placeholder is applied will look like this,

```yaml
apiVersion: scaffolder.backstage.io/v1beta3
kind: Template
metadata:
  name: v1beta3-demo
  title: Microservice example
  description: scaffolder v1beta3 template demo
spec:
  owner: backstage/techdocs-core
  type: service
  parameters:
    - title: Specify a host
      parameters:
        name:
          title: Hostname
          type: string
          description: Specify host name
        domainExt:
          name: Domain extension
          type: string
          description: Specify domain extension like .com, .ca, .co.uk or something else.
```

To implement this, you have to modify the catalog plugin to create a `specifyHostname` resolver. In `/packages/backend/src/plugins/catalog.ts`. Add the following code,

```ts
const builder = await CatalogBuilder.create(env);

builder.setPlaceholderResolver(
    'specifyHostname',
    async specifyHostnameResolver({ value }) => {
      return {
        "title": "Specify a host",
        "parameters": {
            "name": {
                "title": "Hostname",
                "type": "string",
                "description": "Specify host name"
            },
            ...value
        }
      }
    },
  );`
```

### CODE REVIEW

File: plugins/catalog-backend/src/modules/core/PlaceholderProcessor.ts [hunk 1]
```diff
@@ -111,15 +111,19 @@ export class PlaceholderProcessor implements CatalogProcessor {
...
-        // document.
+      if (!resolver) {
+        // If there was no such placeholder resolver, we err on the side of safety
+        // and assume that this is something that's best left alone. For example, if
+        // the input contains JSONSchema, there may be "$ref": "#/definitions/node"
+        // nodes in the document.
         return [data, false];
       }
 
+      // if (typeof resolverValue !== 'string') {
```
Comment (jhaals) [inline plugins/catalog-backend/src/modules/core/PlaceholderProcessor.ts:122]: Some leftover here right?

**Review by jhaals — CHANGES_REQUESTED** [review]
Thanks for the proposal!
Overall I think we should be template other things than string.
However I'm not sure about the documentation section as it's very long which might lead to information overload. This is an advanced use case and I'm not sure that it would be very nice/common to extend templates like this. I would actually suggest that we do not put this example in the docs until we have figured of the patterns for reusable templates.

**Comment by taras** [comment]
> However I'm not sure about the documentation section as it's very long which might lead to information overload. This is an advanced use case and I'm not sure that it would be very nice/common to extend templates like this. I would actually suggest that we do not put this example in the docs until we have figured of the patterns for reusable templates.

Yeah, I can just leave it in this PR as description for someone who wants to try using it.

File: plugins/catalog-backend/src/modules/core/PlaceholderProcessor.ts (hunk 1 above)
Comment (taras) [reply plugins/catalog-backend/src/modules/core/PlaceholderProcessor.ts]: Removed this.

**Comment by taras** [comment]
I removed unnecessary bits.

**Review by jhaals — APPROVED** [review]
Thank you 👍

**Pull request merged**